File: netDxf/IO/DxfRawObjectCommit.cs
// netDxf library. Copyright (c) Daniel Carvajal. Licensed under the MIT License.
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Linq;
using netDxf.Header;

namespace netDxf.IO
{
    public sealed partial class DxfRawObjectTransaction
    {
        /// <summary>Validates and commits all staged edits to a new immutable document. A no-op returns the source instance.</summary>
        /// <remarks>
        /// Failure leaves both source and staging intact so a caller can correct a rejected reference.
        /// Successful commit closes this transaction. No filesystem writes are performed; SaveAtomic is a separate operation.
        /// Only edited schemas are validated; untouched private records remain opaque, not certified valid.
        /// </remarks>
        public DxfRawDocument Commit()
        {
            this.EnsureOpen();
            if (this.undo != null) throw new InvalidOperationException("An object transaction cannot be reentered.");
            if (this.changes.Count == 0 && !this.enableSortents) { this.closed = true; return this.store.Document; }
            DxfRawDocument result = this.Build(true);
            this.ValidateCommit(result);
            this.cancellation.ThrowIfCancellationRequested();
            this.closed = true;
            return result;
        }
        private DxfRawDocument Build(bool metadata)
        {
            DxfRawDocument source = this.store.Document;
            var byStart = new Dictionary<int, Change>();
            foreach (Change change in this.changes.Values) if (change.Source != null) byStart.Add(change.Source.StartTagIndex, change);
            DxfRawSection objects = UniqueSection(source, "OBJECTS");
            var newObjectTags = new List<DxfTag>();
            foreach (ulong key in this.added)
                if (!this.changes[key].Deleted) newObjectTags.AddRange(this.changes[key].Tags);
            var insertions = new Dictionary<int, List<DxfTag>>();
            Action<int, IEnumerable<DxfTag>> insert = (at, tags) =>
            {
                if (!insertions.TryGetValue(at, out List<DxfTag> list)) insertions.Add(at, list = new List<DxfTag>());
                list.AddRange(tags);
            };
            if (objects != null) insert(objects.EndTagIndex - 1, newObjectTags);
            else if (newObjectTags.Count != 0)
            {
                var complete = new List<DxfTag> { new DxfTag(0, "SECTION"), new DxfTag(2, "OBJECTS") };
                complete.AddRange(newObjectTags); complete.Add(new DxfTag(0, "ENDSEC"));
                DxfRawSection thumbnail = UniqueSection(source, "THUMBNAILIMAGE");
                int at = thumbnail == null ? FindEof(source) : thumbnail.StartTagIndex;
                insert(at, complete);
            }
            var tagReplacements = new Dictionary<int, DxfTag>();
            if (metadata)
            {
                this.UpdateHeader(insert, tagReplacements);
                this.UpdateClasses(insert, byStart);
            }
            var result = new List<DxfTag>();
            for (int i = 0; i < source.Tags.Count;)
            {
                if ((i & 1023) == 0) this.cancellation.ThrowIfCancellationRequested();
                if (insertions.TryGetValue(i, out List<DxfTag> insertion)) result.AddRange(insertion);
                if (byStart.TryGetValue(i, out Change replacement))
                {
                    if (!replacement.Deleted) result.AddRange(replacement.Tags);
                    i = replacement.Source.EndTagIndex;
                }
                else
                {
                    result.Add(tagReplacements.TryGetValue(i, out DxfTag replacementTag) ? replacementTag : source.Tags[i]); ++i;
                }
            }
            if (insertions.TryGetValue(source.Tags.Count, out List<DxfTag> trailing)) result.AddRange(trailing);
            this.cancellation.ThrowIfCancellationRequested();
            if (SameTags(source.Tags, result)) return source;
            return source.WithTags(result);
        }
        private static DxfRawSection UniqueSection(DxfRawDocument document, string name)
        {
            DxfRawSection result = null;
            foreach (DxfRawSection section in document.Sections)
                if (StringComparer.OrdinalIgnoreCase.Equals(section.Name, name))
                {
                    if (result != null) throw new InvalidDataException("Multiple " + name + " sections are ambiguous.");
                    result = section;
                }
            return result;
        }
        private static int FindEof(DxfRawDocument source)
        {
            for (int i = source.Tags.Count - 1; i >= 0; --i)
                if (source.Tags[i].Code == 0 && Equals(source.Tags[i].RawValue, "EOF")) return i;
            throw new InvalidDataException("Missing EOF marker.");
        }
        private void UpdateHeader(Action<int, IEnumerable<DxfTag>> insert, Dictionary<int, DxfTag> replacements)
        {
            DxfRawSection header = UniqueSection(this.store.Document, "HEADER");
            if (header == null) throw new NotSupportedException("Object editing requires an explicit HEADER profile.");
            var seeds = header.Records.Where(r => r.Name == "$HANDSEED").ToArray();
            if (seeds.Length > 1) throw new InvalidDataException("Multiple HANDSEED records.");
            if (this.added.Count != 0)
            {
                if (seeds.Length == 0) insert(header.EndTagIndex - 1, new[] { new DxfTag(9, "$HANDSEED"), new DxfTag(5, this.next.ToString("X", CultureInfo.InvariantCulture)) });
                else
                {
                    var seedTags = seeds[0].Tags.Select((t, i) => new { Tag = t, Index = i }).Where(p => p.Tag.Code != 999 && p.Tag.Code != 9).ToArray();
                    if (seedTags.Length != 1 || seedTags[0].Tag.Code != 5) throw new InvalidDataException("Malformed HANDSEED record.");
                    ulong current = DxfObjectText.Number((string)seedTags[0].Tag.RawValue);
                    if (current < this.next) replacements.Add(seeds[0].StartTagIndex + seedTags[0].Index, new DxfTag(5, this.next.ToString("X", CultureInfo.InvariantCulture)));
                }
            }
            if (!this.enableSortents) return;
            var sorting = header.Records.Where(r => r.Name == "$SORTENTS").ToArray();
            if (sorting.Length > 1) throw new InvalidDataException("Multiple SORTENTS variables.");
            if (sorting.Length == 0) insert(header.EndTagIndex - 1, new[] { new DxfTag(9, "$SORTENTS"), new DxfTag(280, (short)16) });
            else
            {
                var flags = sorting[0].Tags.Select((t, i) => new { Tag = t, Index = i }).Where(p => p.Tag.Code != 999 && p.Tag.Code != 9).ToArray();
                if (flags.Length != 1 || flags[0].Tag.Code != 280) throw new InvalidDataException("Malformed SORTENTS variable.");
                short value = (short)flags[0].Tag.RawValue;
                if ((value & 16) == 0) replacements.Add(sorting[0].StartTagIndex + flags[0].Index, new DxfTag(280, (short)(value | 16)));
            }
        }
        private void UpdateClasses(Action<int, IEnumerable<DxfTag>> insert, Dictionary<int, Change> byStart)
        {
            var before = this.store.Objects.GroupBy(o => o.TypeName).ToDictionary(g => g.Key, g => g.Count(), StringComparer.Ordinal);
            var after = this.CurrentObjects().GroupBy(o => o.TypeName).ToDictionary(g => g.Key, g => g.Count(), StringComparer.Ordinal);
            var touched = DxfObjectSchema.RegisteredClasses.Keys.Where(k =>
            { before.TryGetValue(k, out int oldCount); after.TryGetValue(k, out int newCount); return oldCount != newCount; }).OrderBy(x => x, StringComparer.Ordinal).ToArray();
            if (touched.Length == 0) return;
            DxfRawSection classes = UniqueSection(this.store.Document, "CLASSES");
            var missing = new List<DxfTag>();
            foreach (string type in touched)
            {
                this.cancellation.ThrowIfCancellationRequested();
                var found = classes == null ? Array.Empty<DxfRawRecord>() : classes.Records.Where(r => r.Name == "CLASS" && r.Tags.Any(t => t.Code == 1 && Equals(t.RawValue, type))).ToArray();
                if (found.Length > 1) throw new InvalidDataException("Duplicate CLASS definition: " + type);
                after.TryGetValue(type, out int count);
                if (found.Length == 0)
                {
                    if (count == 0) continue;
                    missing.AddRange(new[] { new DxfTag(0, "CLASS"), new DxfTag(1, type),
                        new DxfTag(2, DxfObjectSchema.RegisteredClasses[type]), new DxfTag(3, "ObjectDBX Classes"), new DxfTag(90, 0) });
                    if (this.store.Document.Version >= DxfVersion.AutoCad2004) missing.Add(new DxfTag(91, count));
                    missing.Add(new DxfTag(280, (short)0)); missing.Add(new DxfTag(281, (short)0));
                }
                else
                {
                    DxfRawRecord record = found[0];
                    var names = record.Tags.Where(t => t.Code == 2).ToArray(); var entities = record.Tags.Where(t => t.Code == 281).ToArray();
                    if (names.Length != 1 || !Equals(names[0].RawValue, DxfObjectSchema.RegisteredClasses[type]) ||
                        entities.Length != 1 || (short)entities[0].RawValue != 0)
                        throw new InvalidDataException("Conflicting object CLASS definition: " + type);
                    var counts = record.Tags.Select((t, i) => new { Tag = t, Index = i }).Where(p => p.Tag.Code == 91).ToArray();
                    if (counts.Length > 1) throw new InvalidDataException("Duplicate CLASS instance count.");
                    if (this.store.Document.Version >= DxfVersion.AutoCad2004)
                    {
                        var tags = record.Tags.ToList();
                        if (counts.Length == 1) tags[counts[0].Index] = new DxfTag(91, count);
                        else tags.Insert(tags.FindIndex(t => t.Code == 280) is int position && position >= 0 ? position : tags.Count, new DxfTag(91, count));
                        byStart.Add(record.StartTagIndex, new Change { Source = record, Tags = tags });
                    }
                }
            }
            if (missing.Count == 0) return;
            if (classes != null) insert(classes.EndTagIndex - 1, missing);
            else
            {
                DxfRawSection header = UniqueSection(this.store.Document, "HEADER");
                if (header == null) throw new InvalidDataException("Missing HEADER for CLASSES insertion.");
                var tags = new List<DxfTag> { new DxfTag(0, "SECTION"), new DxfTag(2, "CLASSES") };
                tags.AddRange(missing); tags.Add(new DxfTag(0, "ENDSEC")); insert(header.EndTagIndex, tags);
            }
        }

        private void ValidateCommit(DxfRawDocument document)
        {
            DxfRawObjectStore result = DxfRawObjectStore.Open(document, this.store.Options, this.cancellation);
            foreach (var diagnostic in result.Index.Diagnostics)
                if (diagnostic.Kind == DxfRawHandleDiagnosticKind.DuplicateIdentity || diagnostic.Kind == DxfRawHandleDiagnosticKind.MultipleIdentities ||
                    diagnostic.Kind == DxfRawHandleDiagnosticKind.NullIdentity || diagnostic.Kind == DxfRawHandleDiagnosticKind.InvalidControlGroup ||
                    diagnostic.Kind == DxfRawHandleDiagnosticKind.OwnerCycle || diagnostic.Kind == DxfRawHandleDiagnosticKind.MultipleOwners)
                    throw new InvalidDataException("Committed common graph is invalid: " + diagnostic.Message);
            Action<string> pointer = handle =>
            {
                if (handle == null || handle == "0") return;
                if (result.Index.FindDefinitions(handle).Count != 1) throw new InvalidDataException("Unresolved or ambiguous edited reference: " + handle);
            };
            foreach (var pair in this.changes)
            {
                this.cancellation.ThrowIfCancellationRequested();
                Change change = pair.Value;
                if (change.Deleted || change.Object == null) continue;
                DxfRawStoredObject value = result.Get(change.Object.Handle);
                if (value == null || value is DxfRawOpaqueStoredObject) throw new InvalidDataException("Edited schema could not be read back.");
                pointer(value.OwnerHandle);
                // Schema payload checks below do not cover common controls or appended XData.
                // Validate every interpreted reference on this edited record as well, while
                // leaving opaque application fields and unrelated records untouched.
                foreach (DxfRawHandleOccurrence occurrence in result.Index.GetOccurrences(value.SourceRecord))
                    if (occurrence.IsReference) pointer(occurrence.CanonicalHandle);
                if (value is DxfRawDictionary dictionary)
                {
                    foreach (DxfRawDictionaryEntry entry in dictionary.Entries)
                    {
                        DxfRawStoredObject target = result.Get(entry.Handle) ?? throw new InvalidDataException("Dictionary target is not an OBJECTS record: " + entry.Handle);
                        if (target.OwnerHandle != dictionary.Handle || target.Handle == dictionary.Handle)
                            throw new InvalidDataException("Dictionary entry must target another object with that dictionary as its common owner.");
                    }
                    if (dictionary.HasDefault)
                    {
                        if (dictionary.DefaultHandle == null || dictionary.DefaultHandle == "0" || result.Get(dictionary.DefaultHandle) == null)
                            throw new InvalidDataException("An edited dictionary-with-default requires an existing nonzero default object.");
                    }
                }
                else if (value is DxfRawXRecord record)
                {
                    foreach (DxfTag tag in record.Data)
                        if (tag.Code >= 330 && tag.Code <= 369) pointer(DxfObjectText.Handle((string)tag.RawValue, true));
                }
                else if (value is DxfRawIdBuffer buffer) foreach (string handle in buffer.Handles) pointer(handle);
                else if (value is DxfRawSortentsTable sort)
                {
                    pointer(sort.BlockRecordHandle);
                    if (result.Index.FindDefinitions(sort.BlockRecordHandle)[0].Record.Name != "BLOCK_RECORD")
                        throw new InvalidDataException("Draw-order block reference does not identify a BLOCK_RECORD.");
                    foreach (DxfRawSortOrderEntry entry in sort.Entries) pointer(entry.EntityHandle);
                }
            }
        }
    }
}

File: javascript/netDxf/IO/DxfRawObjectCommit.js
// netDxf library. Copyright (c) Daniel Carvajal. Licensed under the MIT License.
import { DxfRawObjectStore, DxfObjectText as Text, DxfObjectSchema as Schema, ObjectStoreState } from './DxfRawObjectStore.js';
import * as M from './DxfRawObjectModel.js';
import { DxfTag } from './DxfTag.js';
import { EnsureOpen, SameTags, InvalidGraphKinds } from './DxfRawObjectTransaction.js';
import { CurrentObjects } from './DxfRawObjectGraph.js';
import { OrdinalIgnoreCaseEquals as Eq } from '../../runtime/Collections.js';
import * as E from '../../runtime/Errors.js';
const T = (code, value) => new DxfTag(code, value);
const Append = (target, values) => { for (const value of values) target.push(value); };

export function Commit(s) {
  EnsureOpen(s);
  if (s.undo !== null) throw new E.InvalidOperationException('An object transaction cannot be reentered.');
  if (s.changes.size === 0 && !s.enableSortents) { s.closed = true; return s.store.Document; }
  const result = Build(s, true); ValidateCommit(s, result); E.ThrowIfCancellationRequested(s.cancellation);
  s.closed = true; return result;
}
export function Build(s, metadata) {
  const source = s.store.Document, byStart = new Map();
  for (const change of s.changes.values()) if (change.Source !== null) byStart.set(change.Source.StartTagIndex, change);
  const objects = UniqueSection(source, 'OBJECTS'), newObjectTags = [];
  for (const key of s.added) if (!s.changes.get(key).Deleted) Append(newObjectTags, s.changes.get(key).Tags);
  const insertions = new Map();
  const insert = (at, tags) => {
    let list = insertions.get(at); if (!list) insertions.set(at, list = []); Append(list, tags);
  };
  if (objects !== null) insert(objects.EndTagIndex - 1, newObjectTags);
  else if (newObjectTags.length !== 0) {
    const complete = [T(0, 'SECTION'), T(2, 'OBJECTS')]; Append(complete, newObjectTags); complete.push(T(0, 'ENDSEC'));
    const thumbnail = UniqueSection(source, 'THUMBNAILIMAGE'); insert(thumbnail === null ? FindEof(source) : thumbnail.StartTagIndex, complete);
  }
  const replacements = new Map();
  if (metadata) { UpdateHeader(s, insert, replacements); UpdateClasses(s, insert, byStart); }
  const result = [];
  for (let i = 0; i < source.Tags.length;) {
    if ((i & 1023) === 0) E.ThrowIfCancellationRequested(s.cancellation);
    if (insertions.has(i)) Append(result, insertions.get(i));
    const replacement = byStart.get(i);
    if (replacement) {
      if (!replacement.Deleted) Append(result, replacement.Tags);
      i = replacement.Source.EndTagIndex;
    } else { result.push(replacements.get(i) ?? source.Tags[i]); i++; }
  }
  if (insertions.has(source.Tags.length)) Append(result, insertions.get(source.Tags.length));
  E.ThrowIfCancellationRequested(s.cancellation);
  return SameTags(source.Tags, result) ? source : source.WithTags(result);
}
export function UniqueSection(document, name) {
  let result = null;
  for (const section of document.Sections) if (Eq(section.Name, name)) {
    if (result !== null) throw new E.InvalidDataException('Multiple ' + name + ' sections are ambiguous.');
    result = section;
  }
  return result;
}
function FindEof(source) {
  for (let i = source.Tags.length - 1; i >= 0; i--) if (source.Tags[i].Code === 0 && source.Tags[i].Value === 'EOF') return i;
  throw new E.InvalidDataException('Missing EOF marker.');
}
const ValueTags = record => Array.from(record.Tags, (tag, index) => ({ tag, index })).filter(p => p.tag.Code !== 999 && p.tag.Code !== 9);
function UpdateHeader(s, insert, replacements) {
  const header = UniqueSection(s.store.Document, 'HEADER');
  if (header === null) throw new E.NotSupportedException('Object editing requires an explicit HEADER profile.');
  const seeds = header.Records.filter(r => r.Name === '$HANDSEED');
  if (seeds.length > 1) throw new E.InvalidDataException('Multiple HANDSEED records.');
  if (s.added.length !== 0) {
    if (seeds.length === 0) insert(header.EndTagIndex - 1, [T(9, '$HANDSEED'), T(5, s.next.toString(16).toUpperCase())]);
    else {
      const tags = ValueTags(seeds[0]);
      if (tags.length !== 1 || tags[0].tag.Code !== 5) throw new E.InvalidDataException('Malformed HANDSEED record.');
      if (Text.Number(tags[0].tag.Value) < s.next) replacements.set(seeds[0].StartTagIndex + tags[0].index, T(5, s.next.toString(16).toUpperCase()));
    }
  }
  if (!s.enableSortents) return;
  const sorting = header.Records.filter(r => r.Name === '$SORTENTS');
  if (sorting.length > 1) throw new E.InvalidDataException('Multiple SORTENTS variables.');
  if (sorting.length === 0) insert(header.EndTagIndex - 1, [T(9, '$SORTENTS'), T(280, 16)]);
  else {
    const flags = ValueTags(sorting[0]);
    if (flags.length !== 1 || flags[0].tag.Code !== 280) throw new E.InvalidDataException('Malformed SORTENTS variable.');
    const value = flags[0].tag.Value;
    if ((value & 16) === 0) replacements.set(sorting[0].StartTagIndex + flags[0].index, T(280, value | 16));
  }
}
function UpdateClasses(s, insert, byStart) {
  const counts = objects => {
    const result = new Map(); for (const o of objects) result.set(o.TypeName, (result.get(o.TypeName) ?? 0) + 1); return result;
  };
  const before = counts(s.store.Objects), after = counts(CurrentObjects(s));
  const touched = Object.keys(Schema.RegisteredClasses).filter(k => (before.get(k) ?? 0) !== (after.get(k) ?? 0)).sort();
  if (touched.length === 0) return;
  const classes = UniqueSection(s.store.Document, 'CLASSES'), missing = [];
  for (const type of touched) {
    E.ThrowIfCancellationRequested(s.cancellation);
    const found = classes === null ? [] : classes.Records.filter(r => r.Name === 'CLASS' && Array.from(r.Tags).some(t => t.Code === 1 && t.Value === type));
    if (found.length > 1) throw new E.InvalidDataException('Duplicate CLASS definition: ' + type);
    const count = after.get(type) ?? 0;
    if (found.length === 0) {
      if (count === 0) continue;
      missing.push(T(0, 'CLASS'), T(1, type), T(2, Schema.RegisteredClasses[type]), T(3, 'ObjectDBX Classes'), T(90, 0));
      if (s.store.Document.Version >= 14) missing.push(T(91, count));
      missing.push(T(280, 0), T(281, 0));
    } else {
      const record = found[0], tags = Array.from(record.Tags), names = tags.filter(t => t.Code === 2), entities = tags.filter(t => t.Code === 281);
      if (names.length !== 1 || names[0].Value !== Schema.RegisteredClasses[type] || entities.length !== 1 || entities[0].Value !== 0)
        throw new E.InvalidDataException('Conflicting object CLASS definition: ' + type);
      const instanceCounts = tags.map((tag, index) => ({ tag, index })).filter(p => p.tag.Code === 91);
      if (instanceCounts.length > 1) throw new E.InvalidDataException('Duplicate CLASS instance count.');
      if (s.store.Document.Version >= 14) {
        if (instanceCounts.length === 1) tags[instanceCounts[0].index] = T(91, count);
        else { const position = tags.findIndex(t => t.Code === 280); tags.splice(position >= 0 ? position : tags.length, 0, T(91, count)); }
        byStart.set(record.StartTagIndex, { Source: record, Tags: tags });
      }
    }
  }
  if (missing.length === 0) return;
  if (classes !== null) insert(classes.EndTagIndex - 1, missing);
  else {
    const header = UniqueSection(s.store.Document, 'HEADER');
    if (header === null) throw new E.InvalidDataException('Missing HEADER for CLASSES insertion.');
    insert(header.EndTagIndex, [T(0, 'SECTION'), T(2, 'CLASSES'), ...missing, T(0, 'ENDSEC')]);
  }
}
function ValidateCommit(s, document) {
  const result = DxfRawObjectStore.Open(document, s.options, s.cancellation), { index } = ObjectStoreState(result);
  for (const diagnostic of index.Diagnostics) if (InvalidGraphKinds.includes(diagnostic.Kind))
    throw new E.InvalidDataException('Committed common graph is invalid: ' + diagnostic.Message);
  const pointer = handle => {
    if (handle === null || handle === '0') return;
    if (index.FindDefinitions(handle).length !== 1) throw new E.InvalidDataException('Unresolved or ambiguous edited reference: ' + handle);
  };
  for (const change of s.changes.values()) {
    E.ThrowIfCancellationRequested(s.cancellation);
    if (change.Deleted || change.Object === null) continue;
    const value = result.Get(change.Object.Handle);
    if (value === null || value instanceof M.DxfRawOpaqueStoredObject) throw new E.InvalidDataException('Edited schema could not be read back.');
    pointer(value.OwnerHandle);
    for (const occurrence of index.GetOccurrences(value.SourceRecord)) if (occurrence.IsReference) pointer(occurrence.CanonicalHandle);
    if (value instanceof M.DxfRawDictionary) {
      for (const entry of value.Entries) {
        const target = result.Get(entry.Handle);
        if (target === null) throw new E.InvalidDataException('Dictionary target is not an OBJECTS record: ' + entry.Handle);
        if (target.OwnerHandle !== value.Handle || target.Handle === value.Handle)
          throw new E.InvalidDataException('Dictionary entry must target another object with that dictionary as its common owner.');
      }
      if (value.HasDefault && (value.DefaultHandle === null || value.DefaultHandle === '0' || result.Get(value.DefaultHandle) === null))
        throw new E.InvalidDataException('An edited dictionary-with-default requires an existing nonzero default object.');
    } else if (value instanceof M.DxfRawXRecord) {
      for (const tag of value.Data) if (tag.Code >= 330 && tag.Code <= 369) pointer(Text.Handle(tag.Value, true));
    } else if (value instanceof M.DxfRawIdBuffer) for (const handle of value.Handles) pointer(handle);
    else if (value instanceof M.DxfRawSortentsTable) {
      pointer(value.BlockRecordHandle);
      if (index.FindDefinitions(value.BlockRecordHandle)[0].Record.Name !== 'BLOCK_RECORD')
        throw new E.InvalidDataException('Draw-order block reference does not identify a BLOCK_RECORD.');
      for (const entry of value.Entries) pointer(entry.EntityHandle);
    }
  }
}
